File: src/main/java/demos/oop/order/nohiding/OrderLine.java
package demos.oop.order.nohiding;

class OrderLine {
    private String productId;
    private int amount;

    public OrderLine(String productId, int amount) {
        this.productId = productId;
        this.amount = amount;
    }

    public void validate() {
        if (this.amount <= 0 || this.amount > 10) {
            throw new RuntimeException("Order line exceeds the maximum amount");
        }
    }

    public void add(int extraAmount) {
        int newAmount = this.amount + extraAmount;
        if (extraAmount <= 0) {
            throw new RuntimeException("Amount to increase must be positive");
        } else {
            this.amount = newAmount;
            this.validate();
        }
    }

    public String getProductId() {
        return productId;
    }

    public void setProductId(String productId) {
        this.productId = productId;
    }

    public int getAmount() {
        return amount;
    }

    public void setAmount(int amount) {
        this.amount = amount;
    }
}

File: src/main/scala/demos/fp/order/hiding/OrderLine.scala
package demos.fp.order.hiding

class OrderLine(val productId: String, val amount: Int):
    def add(extraAmount: Int): OrderLine =
        val newAmount = this.amount + extraAmount
        if extraAmount <= 0 || newAmount > 10 then throw Exception("Cannot increase order line amount")
        else OrderLine(productId = this.productId, amount = newAmount)

    override def toString: String = s"OrderLine(productId=${this.productId}, amount=${this.amount})"

object OrderLine:
    def apply(productId: String, amount: Int): OrderLine =
        if amount <= 0 || amount > 10 then throw Exception("Cannot create order line")
        else new OrderLine(productId = productId, amount = amount)
